File: 0912-random-pick-with-weight/0912-random-pick-with-weight.cpp
class Solution {
private:
    vector<int> prefixSum;
public:
    Solution(vector<int>& w) {
        for(auto& n: w){
            if(prefixSum.empty()){
                prefixSum.push_back(n);
                continue;
            }
            prefixSum.push_back(n+prefixSum.back());
        }
    }
    
    int pickIndex() {
        float randomNum = (float) rand() / RAND_MAX;
        float targetNum = randomNum * prefixSum.back();
        for(int i = 0; i < prefixSum.size(); i++){
            if(targetNum < prefixSum[i]) return i;
        }

        return prefixSum.size()-1;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */

File: 0912-random-pick-with-weight/0912-random-pick-with-weight.py
class Solution(object):

    def __init__(self, w):
        """
        :type w: List[int]
        """
        self.prefixSum = []
        for n in w:
            if not self.prefixSum:
                self.prefixSum.append(n)
                continue
            self.prefixSum.append(n+self.prefixSum[-1])

    def pickIndex(self):
        """
        :rtype: int
        """
        # random floar between 0 and 9
        randomNum = random.random()
        targetNum = randomNum * self.prefixSum[-1]
        for i in range(len(self.prefixSum)):
            if targetNum < self.prefixSum[i]:
                return i
        
        return len(self.prefxSum)-1
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()